src/cargo/sources/overlay.rs

```diff
@@ -0,0 +1,141 @@
+use std::task::ready;
+
+use crate::sources::IndexSummary;
+
+use super::source::{MaybePackage, Source};
+
+/// A `Source` that overlays a one source over another, pretending that the packages
+/// available in the overlay are actually available in the other one.
+///
+/// This is a massive footgun and a terrible idea, so we do not (and never will)
+/// expose this publicly. However, it is useful for some very specific private
+/// things, like locally verifying a bunch of packages at a time before any of
+/// them have been published.
+pub struct OverlayConfusionAttack<'gctx> {
+    // The overlay source. The naming here comes from the main application of this,
+    // where there is a remote registry that we overlay some local packages on.
+    local: Box<dyn Source + 'gctx>,
+    // The source we're impersonating.
+    remote: Box<dyn Source + 'gctx>,
+}
+
+impl<'gctx> OverlayConfusionAttack<'gctx> {
+    pub fn new(local: Box<dyn Source + 'gctx>, remote: Box<dyn Source + 'gctx>) -> Self {
+        Self { local, remote }
+    }
+}
+
+impl<'gctx> Source for OverlayConfusionAttack<'gctx> {
+    fn source_id(&self) -> crate::core::SourceId {
+        self.remote.source_id()
```

Comment:
In the current implementation that can't happen because the overlays are configured by a `HashMap<SourceId, _>`, so only one overlay can be registered for each remote.